payjoin/src/uri/url_ext.rs

```diff
@@ -173,7 +254,26 @@ mod tests {
         url.set_exp(exp_time);
         assert_eq!(url.fragment(), Some("EX1C4UC6ES"));
 
-        assert_eq!(url.exp(), Some(exp_time));
+        assert_eq!(url.exp().unwrap(), exp_time);
+    }
+
+    #[test]
+    fn test_errors_when_parsing_exp() {
+        let missing_exp_url = Url::parse("http://example.com").unwrap();
+        assert!(matches!(missing_exp_url.exp(), Err(ParseExpParamError::MissingExp)));
+
+        let invalid_bech32_exp_url =
+            Url::parse("http://example.com?pj=https://test-payjoin-url#EX1invalid_bech_32")
+                .unwrap();
+        assert!(matches!(invalid_bech32_exp_url.exp(), Err(ParseExpParamError::DecodeBech32(_))));
+
+        let invalid_hrp_exp_url =
+            Url::parse("http://example.com?pj=https://test-payjoin-url#EX1010").unwrap();
+        assert!(matches!(invalid_hrp_exp_url.exp(), Err(ParseExpParamError::InvalidHrp(_))));
```

Comment:
The param is parsed with the prefix `EX1`, but it grabs everything after that up to the delimiter `+`:
```rust
fn get_param<F, T>(url: &Url, prefix: &str, parse: F) -> Option<T>
where
    F: Fn(&str) -> Option<T>,
{
    if let Some(fragment) = url.fragment() {
        for param in fragment.split('+') { // <--- delimiter specified here
            if param.starts_with(prefix) {
                return parse(param);
            }
        }
    }
    None
}
```

There's a chance that you could get a `1` (bech32 separator) somewhere between the param prefix `EX1` and the delimiter `+`. Since the HRP is everything to the left of the **right-most** separator, the invalid url in the test would have it's HRP being parsed as `EX101` instead of the expected `EX1`.

I pretty much used your implementation for parsing the receiver pubkey for this part and had to make sense of it so it's fresh in my mind lol

Reply:
I agree it's worth the comment. Pushed with the comment added